src/OmniSharp.Roslyn.CSharp/Services/Diagnostics/RoslynAnalyzerService.cs

```diff
@@ -0,0 +1,124 @@
+        private readonly IEnumerable<ICodeActionProvider> providers;
+
+        [ImportingConstructor]
+        public RoslynAnalyzerService(
+            OmniSharpWorkspace workspace, 
+            [ImportMany] IEnumerable<ICodeActionProvider> providers, 
+            ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<RoslynAnalyzerService>();
+            this.providers = providers;
+
+            workspace.WorkspaceChanged += OnWorkspaceChanged;
+
+            Task.Run(() => Worker(CancellationToken.None));
+            Task.Run(() => QueueForAnalysis(workspace.CurrentSolution.Projects));
+        }
+
+        private async Task Worker(CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                var currentWork = GetCurrentWork();
+
+                var analyzerResults = await Task
+                    .WhenAll(currentWork
+                        .Select(async x => new
+                        {
+                            Project = x.Key,
+                            Result = await Analyze(x.Value)
+                        }));
+
+                analyzerResults
+                    .ToList()
+                    .ForEach(result => _results[result.Project] = result.Result);
+
+                await Task.Delay(1000, token);
+            }
+        }
+
+        private ConcurrentDictionary<string, Project> GetCurrentWork()
+        {
+            lock (_workQueue)
+            {
+                var currentWork = _workQueue;
+                _workQueue = new ConcurrentDictionary<string, Project>();
+                return currentWork;
+            }
+        }
+
+        public IEnumerable<DiagnosticLocation> GetCurrentDiagnosticResults() => _results.SelectMany(x => x.Value);
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
+        {
+            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged)
+            {
+                var project = changeEvent.NewSolution.GetDocument(changeEvent.DocumentId).Project;
+                QueueForAnalysis(new[] { changeEvent.NewSolution.GetDocument(changeEvent.DocumentId).Project });
+            }
+        }
+
+        private void QueueForAnalysis(IEnumerable<Project> projects)
+        {
+            projects.ToList().ForEach(project =>
+            {
+                _workQueue.TryAdd(project.Id.ToString(), project);
+            });
+        }
+
+        private async Task<IEnumerable<DiagnosticLocation>> Analyze(Project project)
+        {
+            if (!this.providers.SelectMany(x => x.CodeDiagnosticAnalyzerProviders).Any())
+                return ImmutableArray<DiagnosticLocation>.Empty;
+
+            var compiled = await project.GetCompilationAsync();
+            var analysis = await compiled
+                .WithAnalyzers(this.providers.SelectMany(x => x.CodeDiagnosticAnalyzerProviders).ToImmutableArray())
```

Comment:
A call to `WithAnalyzers` creates a new compilation and so running the analyzers will have to rebind the entire solution - something it would need to do for a regular GetDiagnostics as well. It might be better to compute the analysis results along with the regular compiler diagnostics by calling `CompilationWithAnalyzers.GetAllDiagnosticsAsync()`. That will ensure that the entire solution is bound only once for the diagnostics instead of twice.